File: Source/fireland15.LindenmayerSystem/fireland15.LindenmayerSystem.Expressions/Expression.cs
using fireland15.LindenmayerSystem.Expressions.Model.Nodes;
using System;
using System.Linq;

namespace fireland15.LindenmayerSystem.Expressions.Model
{
    internal class Expression : IExpression
    {
        private readonly ExpressionNode rootNode;

        public IEvaluationContext EvaluationContext { get; }

        public Expression(ExpressionNode rootNode)
        {
            this.rootNode = rootNode;
            EvaluationContext = MakeEvaluationContext(rootNode);
        }
        
        public float Evaluate()
        {
            ValidateContext();
            return rootNode.Evaluate(EvaluationContext);
        }

        private void ValidateContext()
        {
            foreach (var variable in EvaluationContext.Variables)
            {
                if (!variable.IsAssigned)
                    throw new InvalidOperationException($"Variable value unassigned. Variable '{variable.Symbol}' was not assigned a value.");
            }
        }

        private IEvaluationContext MakeEvaluationContext(ExpressionNode rootNode) 
        {
            var variableMap = rootNode.Descendants
                .Append(rootNode)
                .OfType<VariableNode>()
                .Select(x => x.Symbol)
                .ToDictionary(
                    symbol => symbol,
                    symbol => new Variable(symbol));

            return new EvaluationContext(variableMap);
        }
    }
}

File: Source/archive/expression-evaluator/expression.cpp
#include "expression.h"

Expression::Expression(ExpressionNode* expressionTree, std::map<std::string, std::list<VariableNode*>> variableMap)
	: m_expressionRoot(expressionTree), m_variableMap(variableMap) { }

float Expression::Evaluate() {
	m_result = (*m_expressionRoot)();
	return m_result;
}

void Expression::SetVariable(const std::string& symbol, float value) {
	if (m_variableMap.find(symbol) != m_variableMap.end()) {
		for (auto it = m_variableMap[symbol].begin(); it != m_variableMap[symbol].end(); ++it) {
			(*it)->SetValue(value);
		}
	}
}

float Expression::GetResult() {
	return m_result;
}